samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManagerUtil.java

```diff
@@ -411,4 +438,292 @@ public static Set<String> getSideInputStoreNames(
+      ExecutorService executor, Set<String> storesToRestore, File loggedStoreBaseDirectory, JobContext jobContext,
+      ContainerModel containerModel) {
+
+    // if taskInstanceMetrics are specified use those for store metrics,
+    // otherwise (in case of StorageRecovery) use a blank MetricsRegistryMap
+    MetricsRegistry metricsRegistry =
+        taskInstanceMetrics.get(taskName) != null ? taskInstanceMetrics.get(taskName).registry()
+            : new MetricsRegistryMap();
+
+    BlobStoreManager blobStoreManager = getBlobStoreManager(config, executor);
+    JobConfig jobConfig = new JobConfig(config);
+    BlobStoreUtil blobStoreUtil =
+        new BlobStoreUtil(blobStoreManager, executor, new BlobStoreConfig(config), null,
+            new BlobStoreRestoreManagerMetrics(metricsRegistry));
+
+    BlobStoreRestoreManager blobStoreRestoreManager = (BlobStoreRestoreManager) taskRestoreManager;
+
+    CheckpointId checkpointId = CheckpointId.create();
+    Map<String, String> oldSCMs = ((CheckpointV2) taskCheckpoints.get(taskName))
+        .getStateCheckpointMarkers().get(BlobStoreStateBackendFactory.class.getName());
+
+    // Returns a single future that guarantees all the following are completed, in this order:
+    // 1. Restore state locally by getting deleted blobs from the blob store.
+    // 2. Create a new Checkpoint from restored state and back it up on the blob store.
+    // 3. Clean up old/deleted Snapshot
+    // 4. Remove TTL from the new Snapshot on the blob store
+    // 5. Write the new checkpoint to checkpoint topic and update taskCheckpoints map
+
+    // 1. restore state with getDeleted flag set to true
+    return blobStoreRestoreManager.restore(true).thenCompose(r -> {
+      // 2. Create and backup new checkpoint on the blob store.
+      CompletableFuture<Map<String, String>> uploadSCMs =
+          createNewCheckpointAndBackupStores(jobContext, containerModel, config, taskName, storesToRestore, checkpointId,
+              loggedStoreBaseDirectory, blobStoreManager, metricsRegistry, executor);
+      CompletableFuture<Void> future = uploadSCMs.thenCompose(scms -> {
+        // 3. Delete prev SnapshotIndex including files/subdirs,
+        Map<String, CompletionStage<Void>> deleteOldSnapshotsFutures =
+            deletePrevSnapshots(blobStoreUtil, oldSCMs, taskName, jobConfig);
+        // 4. Mark new Snapshots to never expire
+        Map<String, CompletionStage<Void>> removeNewSnapshotsTTLFutures =
+            removeTTLNewSnapshots(blobStoreUtil, scms, taskName, jobConfig);
+        // Combined future of delete old snapshots and removeTTL of new snapshots
+        CompletableFuture<Void> removeTTLFuture = CompletableFuture.allOf(FutureUtil.mapToFuture(deleteOldSnapshotsFutures),
+            FutureUtil.mapToFuture(removeNewSnapshotsTTLFutures));
+        // 5. Update taskCheckpoints map with new checkpoint
+        CompletableFuture<Void> checkpointFuture = removeTTLFuture.thenRun(() -> {
```

Comment:
should this be andThen instead of thenRun? thenRun is fire and forget and is not guaranteed to complete when this future completes.

Generally, should not thenCompose recursively in error handlers, e.g. line 602. Chain on the original future to clarify execution DAG.

Also, do you need to specify the executor these operations are happening on?

Reply:
Updated to clarify each step and their dependency.